### ISSUE TITLE
Question about lovasz_hinge_loss

### ISSUE THREAD

**Issue by Epiphqny** [yinghdb/EmbedMask#14]
What's the purpose of act in https://github.com/yinghdb/EmbedMask/blob/3b4d9fb4e0b6112dc501708184ff684dfb45f3f0/fcos_core/layers/lovasz_hinge_loss.py#L64, i find there is no act in the original implementation of lovasz_hinge. Turning on it could cause nan in the training process, but turning off it the loss does not degrad. Do you have any idea about the nan problem, what might be the reason? Reducing the lr does not help.

**Comment by yinghdb** [comment]
Right. The original lovasz_hinge does not have the 'act' operation. But the input range of the original lovasz_hinge is (-inf, +inf). So I use the 'act' operation to range my input. And it brings 'nan' problem. 

The 'nan' problem mainly come from the extreme value of the input, e.g., the original input which is very close to 0 will be come the nearly '-inf' after 'act'. In my later practice, I find it useful to eliminate these extreme values before 'act' like following, and it rarely affects the final results. 

```
m = (input_flatten < 0.999999) * (input_flatten > 1e-9)  
input_flatten = input_flatten[m]  
target_flatten = target_flatten[m]
input_flatten = torch.log(input_flatten) - torch.log(1 - input_flatten)
```

I am struggling in improving the performance of EmbedMask, so I have not fixed the problems like this, sorry!

**Comment by Epiphqny** [comment]
Thanks for your quick reply, i will try the solution. Hope to see the improvements of EmbedMask!

**Comment by Yuxin-CV** [comment]
> Right. The original lovasz_hinge does not have the 'act' operation. But the input range of the original lovasz_hinge is (-inf, +inf). So I use the 'act' operation to range my input. And it brings 'nan' problem.
> 
> The 'nan' problem mainly come from the extreme value of the input, e.g., the original input which is very close to 0 will be come the nearly '-inf' after 'act'. In my later practice, I find it useful to eliminate these extreme values before 'act' like following, and it rarely affects the final results.
> 
> ```
> m = (input_flatten < 0.999999) * (input_flatten > 1e-9)  
> input_flatten = input_flatten[m]  
> target_flatten = target_flatten[m]
> input_flatten = torch.log(input_flatten) - torch.log(1 - input_flatten)
> ```
> 
> I am struggling in improving the performance of EmbedMask, so I have not fixed the problems like this, sorry!

I think it might be better to use `torch.clamp()`.
```
if act:
    # map [0, 1] to [-inf, inf]

    # m = (input_flatten < 0.999999) * (input_flatten > 1e-9) 
    # input_flatten = input_flatten[m]  
    # target_flatten = target_flatten[m]

    MIN = 1e-6
    MAX = 1 - MIN

    torch.clamp(input=input_flatten,  min=MIN, max=MAX, out=input_flatten)
    torch.clamp(input=target_flatten, min=MIN, max=MAX, out=target_flatten)

    input_flatten = torch.log(input_flatten) - torch.log(1 - input_flatten) # prob to logits
```
Since `m = (input_flatten < 0.999999) * (input_flatten > 1e-9)`  reduces the number of samples & targets.
